crosscluster: monitor lagging spans

msbutler opened the pull request:
This patch teaches the ldr to collect and aggregration the count of source side ranges undergoing catchup and initial scans. In addition this patch reports this information in the job's running status and fraction completed.

Epic: none

Release note: none

jeffswenson committed:
crosscluster: monitor lagging spans

This patch teaches the ldr to collect and aggregration the count of source side
ranges undergoing catchup and initial scans. In addition this patch reports
this information in the job's running status and fraction completed.

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134090)
<!-- Reviewable:end -->

msbutler committed:
crosscluster/logical: add initial and catchup scan metrics

Epic: none

Release note: this patch adds the following LDR metrics
 - logical_replication.catchup_ranges: the number of source side ranges
   conducting catchup scans.
 -logical_replication.scanning_ranges: the number source side ranges conducting
  initial scans.
Note that in the dbconsole, these metrics are not accurate if multiple LDR jobs
are running, though there exists the equivalent labeled metrics for a user to
consume via prometheus.

msbutler commented:
Initial scan on job page:
<img width="425" alt="image" src="https://github.com/user-attachments/assets/a67a2010-eabb-4bc0-a771-4e8a3e8f807e">

All caught up:
<img width="407" alt="image" src="https://github.com/user-attachments/assets/9f6c5362-20ba-4880-b06c-1be40ef651a9">

Just after a resume:
<img width="411" alt="image" src="https://github.com/user-attachments/assets/95a556cd-8c53-4f5f-b56b-95e352468b00">

Catchup scans:
<img width="397" alt="image" src="https://github.com/user-attachments/assets/e80dad6d-e660-453f-b4db-b2b599190524">

msbutler commented:
Scanning and Lagging ranges metrics, because the source only polls once a minute, any initial or catchup scan that takes less than a minute (like this one) will look slightly off.
<img width="996" alt="image" src="https://github.com/user-attachments/assets/47d50ae1-a6ce-461d-9095-386e2823d7f4">

msbutler commented:
unrelated unit test flake

jeffswenson commented on pkg/ccl/crosscluster/logical/range_stats.go:
```diff
@@ -0,0 +1,63 @@
+)
+
+type rangeStatsByProcessorID struct {
+	mu             syncutil.Mutex
```
Is the mutex required here? Are rows and producer metas produced in parallel?

msbutler replied:
i don't know if they can produced in parallel, but i thought better safe than sorry, especially because these apis aren't called too often.

jeffswenson commented on pkg/sql/execinfrapb/data.proto:
```diff
@@ -329,6 +329,8 @@ message RemoteProducerMetadata {
       (gogoproto.customname) = "FlowID",
       (gogoproto.customtype) = "FlowID"];
     optional bool drained = 9 [(gogoproto.nullable) = false];
+		// ProcessorID is the ID of the processor that published the metadata.
```
nit / (why isn't there a formatter for this 😢 ): this comment is indented with tabs when it should be spaces

msbutler replied:
i can't tell you how many times i've tried to coax vscode to use spaces instead of tabs.

jeffswenson commented on pkg/ccl/crosscluster/logical/logical_replication_job.go:
```diff
@@ -532,13 +552,31 @@ func (rh *rowHandler) handleRow(ctx context.Context, row tree.Datums) error {
 			}
-			progress.RunningStatus = fmt.Sprintf("logical replication running: %s", replicatedTime.GoTime())
+			progress.RunningStatus = status
+			if fractionCompleted > 0 {
```
I think we only want to show the progress bar if 0 < fractionCompleted < 1. Currently, this is overwriting the high watermark, so we will no longer show the high watermark when everything is caught up and advancing.

msbutler replied:
I like that idea

msbutler replied:
if fraction completed is 0, the status is now "all ranges are caught up"

jeffswenson reviewed (approved):
LGTM

msbutler commented:
TFTR!

bors r=dt
